zict/sieve.py

```diff
@@ -36,63 +37,83 @@ class Sieve(ZictBase[KT, VT], Generic[KT, VT, MKT]):
     >>> def is_small(key, value):                 # doctest: +SKIP
     ...     return sys.getsizeof(value) < 10000   # doctest: +SKIP
     >>> d = Sieve(mappings, is_small)             # doctest: +SKIP
-
-    See Also
-    --------
-    Buffer
     """
 
     mappings: Mapping[MKT, MutableMapping[KT, VT]]
     selector: Callable[[KT, VT], MKT]
     key_to_mapping: dict[KT, MutableMapping[KT, VT]]
+    gen: int
 
     def __init__(
         self,
         mappings: Mapping[MKT, MutableMapping[KT, VT]],
         selector: Callable[[KT, VT], MKT],
     ):
+        super().__init__()
         self.mappings = mappings
         self.selector = selector
         self.key_to_mapping = {}
+        self.gen = 0
 
     def __getitem__(self, key: KT) -> VT:
+        # Note that this may raise KeyError if you call it in the middle of __setitem__
+        # or update for an already existing key
         return self.key_to_mapping[key][key]
 
+    @locked
     def __setitem__(self, key: KT, value: VT) -> None:
-        old_mapping = self.key_to_mapping.get(key)
+        discard(self, key)
         mkey = self.selector(key, value)
         mapping = self.mappings[mkey]
-        if old_mapping is not None and old_mapping is not mapping:
-            del old_mapping[key]
-        mapping[key] = value
         self.key_to_mapping[key] = mapping
+        self.gen += 1
+        gen = self.gen
+
+        with self.unlock():
+            mapping[key] = value
 
+        if gen != self.gen and self.key_to_mapping.get(key) is not mapping:
+            # Multithreaded race condition
+            discard(mapping, key)
```

Comment:
I suppose this is a similar clarification; is the tradeoff  worth potentially corrupting the expected state and trying to rectify it?